test/drivers/test_keysight_33500b_pytest.py

```diff
@@ -0,0 +1,448 @@
+from fixate.config import load_config
+load_config() # Load fixate config file
+
+# rewirte of "test_keysight_33500b.py" into pytest framework
+
+# TODO: Use fixture setup and taredown to clean up test code.
+# TODO: Separate early tests with multiple asserts into tests with one assert each. Make use of mark parameterised
+
+
+@pytest.mark.drivertest
+def test_open_funcgen():
+    import fixate.drivers.funcgen
+    funcgen = fixate.drivers.funcgen.open()
+    assert funcgen, "Could not open function generator"
+
+@pytest.mark.drivertest
+def test_sin(funcgen):
+    funcgen.channel1.waveform.square()
+    funcgen.channel1.waveform.sin()
+    assert "SIN\n" == funcgen.instrument.query("SOUR1:FUNC?")
+
+@pytest.mark.drivertest
+def test_square(funcgen):
+    funcgen.channel1.waveform.square()
+    assert "SQU\n" == funcgen.instrument.query("SOUR1:FUNC?")
+
+@pytest.mark.drivertest
+def test_ramp(funcgen):
+    funcgen.channel1.waveform.ramp()
+    assert "RAMP\n" == funcgen.instrument.query("SOUR1:FUNC?")
+
+@pytest.mark.drivertest
+def test_pulse(funcgen):
+    funcgen.channel1.waveform.pulse()
+    assert "PULS\n" == funcgen.instrument.query("SOUR1:FUNC?")
+
+@pytest.mark.drivertest
+def test_arb(funcgen):
+    funcgen.channel1.waveform.arb()
+    assert "ARB\n" == funcgen.instrument.query("SOUR1:FUNC?")
+
+@pytest.mark.drivertest
+def test_triangle(funcgen):
+    funcgen.channel1.waveform.triangle()
+    assert "TRI\n" == funcgen.instrument.query("SOUR1:FUNC?")
+
+@pytest.mark.drivertest
+def test_noise(funcgen):
+    funcgen.channel1.vrms(0.2) # Nosie has a small VRMS limit. Set this to not error
+    funcgen.channel1.waveform.noise()
+    assert "NOIS\n" == funcgen.instrument.query("SOUR1:FUNC?")
+
+@pytest.mark.drivertest
+def test_dc(funcgen):
+    funcgen.channel1.waveform.dc()
+    assert "DC\n" == funcgen.instrument.query("SOUR1:FUNC?")
+
+@pytest.mark.drivertest
+def test_prbs(funcgen):
+    funcgen.channel1.waveform.prbs()
+    assert "PRBS\n" == funcgen.instrument.query("SOUR1:FUNC?")
+
+@pytest.mark.drivertest
+def test_vrms(funcgen):
+    funcgen.reset()
+    funcgen.channel1.load(50)
+    funcgen.channel1.vrms(2)
+
+    # Units
+    assert "VRMS\n" == funcgen.instrument.query("SOUR1:VOLT:UNIT?")
+    
+    # Nominal Level
+    assert float(funcgen.instrument.query("SOUR1:VOLT?")) == pytest.approx(2.0)
+
+    # Upper Limits
+    funcgen.channel1.vrms(3.5)
+    assert float(funcgen.instrument.query("SOUR1:VOLT?")) == pytest.approx(3.5)
+    with pytest.raises(InstrumentError) as excinfo:
+        funcgen.channel1.vrms(3.6)
+
+    assert re.search("value clipped to upper limit", str(excinfo.value))
+
+    # Lower Limits
+    funcgen.channel1.vrms(354e-6)
+    assert float(funcgen.instrument.query("SOUR1:VOLT?")) == pytest.approx(354e-6)
+    with pytest.raises(InstrumentError) as excinfo:
+        funcgen.channel1.vrms(353e-6)
+
+    assert re.search("value clipped to lower limit", str(excinfo.value))
```

Comment:
Should break these up as individual tests and parameterise each of the functions and input values